### PR TITLE
feat(spin): Adds `fullscreen` property to `<Spin />` component

### LINKED ISSUE #44829: Backdrop component
### What problem does this feature solve?

Does antd have a component like mui Backdrop component which can be used as the suspense's fallback loading component.

### What does the proposed API look like?

https://mui.com/material-ui/api/backdrop/

### PR DESCRIPTION
**Pull request by Rafael-Martins**
### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [x] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

- close #44829 

### 💡 Background and solution

Create a solution similar to https://mui.com/material-ui/api/backdrop/ but inside the `<Spin />` component

### 📝 Changelog

Added a new prop called `fullscreen` that adds a css class to make it behave as it should.

### Video

https://github.com/ant-design/ant-design/assets/27746333/57d8d63c-dbcb-4355-9e47-cabc42afc1cf

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |          |
| 🇨🇳 Chinese |           |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at b2fa189</samp>

This pull request adds a new feature to the `Spin` component that allows it to display a fullscreen spinner and handle click events. It also adds a new demo, documentation, and style for the feature. The files affected are `components/spin/index.tsx`, `components/spin/style/index.tsx`, `components/spin/demo/fullscreen.md`, `components/spin/demo/fullscreen.tsx`, and `components/spin/index.en-US.md`.

### <samp>🤖 Generated by Copilot at b2fa189</samp>

*  Add a new fullscreen demo for the Spin component ([link](https://github.com/ant-design/ant-design/pull/44986/files?diff=unified&w=0#diff-123bbad5a2943dfbd8a917c505db96dd1083cc2fbca04d63ba89f1c4eefe976eR1-R7), [link](https://github.com/ant-design/ant-design/pull/44986/files?diff=unified&w=0#diff-27e2c3ba091ddbe9526600f63ec3dea9d054923496acd3f4fc8dccb34d7359f8R1-R22), [link](https://github.com/ant-design/ant-design/pull/44986/files?diff=unified&w=0#diff-24e1b69a8eb3c64c3011f19e7d226ce549861b3e27f628c74ee86f8464c5ff9aL20-R27))
*  Add new props `fullscreen` and `onClick` to the Spin component ([link](https://github.com/ant-design/ant-design/pull/44986/files?diff=unified&w=0#diff-5703e1f246d34eb8b38fcf640e6b64b44e3a1131cdbd49306153af3c1bc0618aR28-R29), [link](https://github.com/ant-design/ant-design/pull/44986/files?diff=unified&w=0#diff-5703e1f246d34eb8b38fcf640e6b64b44e3a1131cdbd49306153af3c1bc0618aR92-R93), [link](https://github.com/ant-design/ant-design/pull/44986/files?diff=unified&w=0#diff-5703e1f246d34eb8b38fcf640e6b64b44e3a1131cdbd49306153af3c1bc0618aR133), [link](https://github.com/ant-design/ant-design/pull/44986/files?diff=unified&w=0#diff-5703e1f246d34eb8b38fcf640e6b64b44e3a1131cdbd49306153af3c1bc0618aR157))
*  Add style definition for the fullscreen spinner in `components/spin/style/index.tsx` ([link](https://github.com/ant-design/ant-design/pull/44986/files?diff=unified&w=0#diff-3a2474e4dab48a16cce22c61f21d534393edcd5b4131bb0089537002750eba24R3), [link](https://github.com/ant-design/ant-design/pull/44986/files?diff=unified&w=0#diff-3a2474e4dab48a16cce22c61f21d534393edcd5b4131bb0089537002750eba24R61-R73))

### CODE REVIEW

**Comment by li-jia-nan** [comment]
PR to feature branch please.

File: components/spin/index.en-US.md [hunk 1]
```diff
@@ -37,6 +38,7 @@ Common props ref：[Common props](/docs/react/common-props)
 | spinning | Whether Spin is visible | boolean | true |
 | tip | Customize description content when Spin has children | ReactNode | - |
 | wrapperClassName | The className of wrapper when Spin has children | string | - |
+| fullscreen | Display a backdrop with the `Spin` component | boolean | false |
```
Comment (li-jia-nan) [inline components/spin/index.en-US.md:41]: You need to add the api version:

```suggestion
| fullscreen | Display a backdrop with the `Spin` component | boolean | false | 5.10.0 |
```

File: components/spin/index.zh-CN.md [hunk 2]
```diff
@@ -38,6 +39,7 @@ demo:
 | spinning         | 是否为加载中状态                             | boolean       | true      |
 | tip              | 当作为包裹元素时，可以自定义描述文案         | ReactNode     | -         |
 | wrapperClassName | 包装器的类属性                               | string        | -         |
+| fullscreen       | 显示带有`Spin`组件的背景                     | boolean       | false     |
```
Comment (li-jia-nan) [inline components/spin/index.zh-CN.md:42]: You need to add the api version:

```suggestion
| fullscreen       | 显示带有 `Spin` 组件的背景                     | boolean       | false     | 5.10.0 |
```

File: components/spin/style/index.tsx [hunk 3]
```diff
@@ -57,6 +58,19 @@ const genSpinStyle: GenerateStyle<SpinToken> = (token: SpinToken): CSSObject =>
       opacity: 1,
     },
 
+    '&-fullscreen': {
+      position: 'fixed',
+      width: '100vw',
+      height: '100vh',
+      background: 'rgba(0,0,0,0.2)',
+      zIndex: 1000,
```
Comment (li-jia-nan) [inline components/spin/style/index.tsx:66]:
```suggestion
      zIndex: token.zIndexPopupBase,
```

File: components/spin/index.en-US.md [hunk 4]
```diff
@@ -17,26 +17,28 @@ When part of the page is waiting for asynchronous data or during a rendering pro
...
-<code src="./demo/delayAndDebounce.tsx">delay</code>
+<code src="./demo/delayAndDebounce.tsx">Delay</code>
 <code src="./demo/custom-indicator.tsx">Custom spinning indicator</code>
+<code src="./demo/fullscreen.tsx">Fullscreen</code>
 
 ## API
 
 Common props ref：[Common props](/docs/react/common-props)
 
 | Property | Description | Type | Default |
```
Comment (li-jia-nan) [inline components/spin/index.en-US.md:33]:
```suggestion
| Property | Description | Type | Default | Version |
```

File: components/spin/index.zh-CN.md [hunk 5]
```diff
@@ -25,19 +25,21 @@ demo:
 <code src="./demo/tip.tsx">自定义描述文案</code>
 <code src="./demo/delayAndDebounce.tsx">延迟</code>
 <code src="./demo/custom-indicator.tsx">自定义指示符</code>
+<code src="./demo/fullscreen.tsx">全屏</code>
 
 ## API
 
 通用属性参考：[通用属性](/docs/react/common-props)
 
 | 参数             | 说明                                         | 类型          | 默认值    |
```
Comment (li-jia-nan) [inline components/spin/index.zh-CN.md:34]:
```suggestion
| 参数             | 说明                                         | 类型          | 默认值    | 版本 |
```

File: components/spin/style/index.tsx [hunk 6]
```diff
@@ -57,6 +58,19 @@ const genSpinStyle: GenerateStyle<SpinToken> = (token: SpinToken): CSSObject =>
       opacity: 1,
     },
 
+    '&-fullscreen': {
+      position: 'fixed',
+      width: '100vw',
+      height: '100vh',
+      background: 'rgba(0,0,0,0.2)',
+      zIndex: token.zIndexPopupBase,
+      top: 0,
```
Comment (li-jia-nan) [inline components/spin/style/index.tsx:67]: You can replace `top` `left` `right` `button` with  `inset: 0`

File: components/spin/index.tsx [hunk 7]
```diff
@@ -126,6 +130,7 @@ const Spin: React.FC<SpinClassProps> = (props) => {
       [`${prefixCls}-lg`]: size === 'large',
       [`${prefixCls}-spinning`]: spinning,
       [`${prefixCls}-show-text`]: !!tip,
+      [`${prefixCls}-fullscreen`]: !!fullscreen,
```
Comment (li-jia-nan) [inline components/spin/index.tsx:133]:
```suggestion
      [`${prefixCls}-fullscreen`]: fullscreen,
```

Comment (Rafael-Martins) [reply components/spin/index.tsx:133]: it's already boolean hahaha :man_facepalming:

File: components/spin/style/index.tsx [hunk 8]
```diff
@@ -57,6 +58,18 @@ const genSpinStyle: GenerateStyle<SpinToken> = (token: SpinToken): CSSObject =>
       opacity: 1,
     },
 
+    '&-fullscreen': {
+      position: 'fixed',
+      width: '100vw',
+      height: '100vh',
+      background: 'rgba(0,0,0,0.2)',
```
Comment (li-jia-nan) [inline components/spin/style/index.tsx:65]:
```suggestion
      backgroundColor: token.colorBgMask,
```

File: components/spin/index.tsx [hunk 9]
```diff
@@ -25,6 +25,8 @@ export interface SpinProps {
   wrapperClassName?: string;
   indicator?: SpinIndicator;
   children?: React.ReactNode;
+  fullscreen?: boolean;
+  onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
```
Comment (li-jia-nan) [inline components/spin/index.tsx:29]: I don't think we need `onClick` event

Comment (Rafael-Martins) [reply components/spin/index.tsx:29]: I kinda agree, I just added to match the material-ui approach, I will remove and change the way that demo works!

Comment (li-jia-nan) [reply components/spin/index.tsx:29]: It can be kept temporarily. Regarding onClick, this still needs to be discussed by the antd team.

Comment (li-jia-nan) [reply components/spin/index.tsx:29]: @MadCcc @zombieJ Do you think `onClick` should be retained?

File: components/spin/style/index.tsx (hunk 8 above)
Comment (Rafael-Martins) [reply components/spin/style/index.tsx:65]: I thought in use that, but it feel a little too much dark, do we have a lighter one? or, can I create a component token?

image

Comment (li-jia-nan) [reply components/spin/style/index.tsx:65]: Yes, you can create custom token, but this is not the final style, we still need to wait for the discussion results from the designers.

File: components/spin/index.tsx (hunk 9 above)
Comment (Rafael-Martins) [reply components/spin/index.tsx:29]: All right!, I will add to the `onClick` to the docs if gonna kept it.

Comment (zombieJ) [reply components/spin/index.tsx:29]: In my view, Spin should not provide `onClick` since we do not expect user to interactive with this.

**Comment by Rafael-Martins** [comment]
I automated the close action in the demo with 2s:

https://github.com/ant-design/ant-design/assets/27746333/1b42a703-130a-478d-a93e-e5fba7a161db

We can also set a close button instead, not sure what is the best approach :thinking:

**Comment by Brucia323** [comment]
I don’t know if Gaussian blur will look good, `backdrop-filter: blur(8px);`

**Comment by li-jia-nan** [comment]
Designer’s discussion results: Change dots to white 85%、65%、45%、35%，and mask backgroundColor reuse modal mask.

File: components/spin/index.en-US.md (hunk 1 above)
Comment (yoyo837) [reply components/spin/index.en-US.md]: Now it seems that it should be landed in version 5.11.0.

File: components/spin/style/index.tsx [hunk 10]
```diff
@@ -57,6 +58,21 @@ const genSpinStyle: GenerateStyle<SpinToken> = (token: SpinToken): CSSObject =>
...
+      width: '100vw',
+      height: '100vh',
+      backgroundColor: token.colorBgMask,
+      zIndex: token.zIndexPopupBase,
+      inset: 0,
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'center',
+      [`${token.componentCls}-dot ${token.componentCls}-dot-item`]: {
+        backgroundColor: '#fff',
```
Comment (li-jia-nan) [inline components/spin/style/index.tsx:72]: The white here should use token instead of constant

File: components/spin/style/index.tsx [hunk 11]
```diff
@@ -57,6 +58,21 @@ const genSpinStyle: GenerateStyle<SpinToken> = (token: SpinToken): CSSObject =>
...
+      width: '100vw',
+      height: '100vh',
+      backgroundColor: token.colorBgMask,
+      zIndex: token.zIndexPopupBase,
+      inset: 0,
+      display: 'flex',
+      alignItems: 'center',
+      justifyContent: 'center',
+      [`${token.componentCls}-dot ${token.componentCls}-dot-item`]: {
+        backgroundColor: `${token.colorWhite}`,
```
Comment (li-jia-nan) [inline components/spin/style/index.tsx:72]:
```suggestion
        backgroundColor: token.colorWhite,
```

File: components/spin/index.en-US.md [hunk 12]
```diff
@@ -17,26 +17,28 @@ When part of the page is waiting for asynchronous data or during a rendering pro
...
-| --- | --- | --- | --- |
+| Property | Description | Type | Default | Version |
+| --- | --- | --- | --- | --- |
 | delay | Specifies a delay in milliseconds for loading state (prevent flush) | number (milliseconds) | - |
 | indicator | React node of the spinning indicator | ReactNode | - |
 | size | The size of Spin, options: `small`, `default` and `large` | string | `default` |
 | spinning | Whether Spin is visible | boolean | true |
 | tip | Customize description content when Spin has children | ReactNode | - |
 | wrapperClassName | The className of wrapper when Spin has children | string | - |
+| fullscreen | Display a backdrop with the `Spin` component | boolean | false | 5.10.0 |
```
Comment (li-jia-nan) [inline components/spin/index.en-US.md:41]:
```suggestion
| fullscreen | Display a backdrop with the `Spin` component | boolean | false | 5.11.0 |
```

File: components/spin/index.zh-CN.md [hunk 13]
```diff
@@ -25,19 +25,21 @@ demo:
...
-| wrapperClassName | 包装器的类属性                               | string        | -         |
+| 参数 | 说明 | 类型 | 默认值 | 版本 |
+| --- | --- | --- | --- | --- |
+| delay | 延迟显示加载效果的时间（防止闪烁） | number (毫秒) | - |
+| indicator | 加载指示符 | ReactNode | - |
+| size | 组件大小，可选值为 `small` `default` `large` | string | `default` |
+| spinning | 是否为加载中状态 | boolean | true |
+| tip | 当作为包裹元素时，可以自定义描述文案 | ReactNode | - |
+| wrapperClassName | 包装器的类属性 | string | - |
+| fullscreen | 显示带有`Spin`组件的背景 | boolean | false | 5.10.0 |
```
Comment (li-jia-nan) [inline components/spin/index.zh-CN.md:42]:
```suggestion
| fullscreen | 显示带有`Spin`组件的背景 | boolean | false | 5.11.0 |
```

File: components/spin/index.zh-CN.md [hunk 14]
```diff
@@ -25,19 +25,21 @@ demo:
...
-| wrapperClassName | 包装器的类属性                               | string        | -         |
+| 参数 | 说明 | 类型 | 默认值 | 版本 |
+| --- | --- | --- | --- | --- |
+| delay | 延迟显示加载效果的时间（防止闪烁） | number (毫秒) | - |
+| indicator | 加载指示符 | ReactNode | - |
+| size | 组件大小，可选值为 `small` `default` `large` | string | `default` |
+| spinning | 是否为加载中状态 | boolean | true |
+| tip | 当作为包裹元素时，可以自定义描述文案 | ReactNode | - |
+| wrapperClassName | 包装器的类属性 | string | - |
+| fullscreen | 显示带有`Spin`组件的背景 | boolean | false | 5.11.0 |
```
Comment (MadCcc) [inline components/spin/index.zh-CN.md:42]:
```suggestion
| fullscreen | 显示带有 `Spin` 组件的背景 | boolean | false | 5.11.0 |
```

File: components/spin/demo/fullscreen.md [hunk 15]
```diff
@@ -0,0 +1,7 @@
+## zh-CN
+
+“全屏”属性非常适合创建流畅的页面加载器。它添加了半透明覆盖层，并在其中心放置了一个旋转加载符号。
```
Comment (MadCcc) [inline components/spin/demo/fullscreen.md:3]:
```suggestion
`fullscreen` 属性非常适合创建流畅的页面加载器。它添加了半透明覆盖层，并在其中心放置了一个旋转加载符号。
```

**Comment by MadCcc** [comment]
`tip` should be enabled for `fullscreen`

**Comment by Rafael-Martins** [comment]
Hey @MadCcc, I fixed the tip usage, I just feel that I'm repeating code a little bit in css for this, what do you think?

https://github.com/ant-design/ant-design/assets/27746333/bf84e82a-4759-4134-b2bc-a3599377aea6

File: components/spin/index.tsx [hunk 16]
```diff
@@ -155,7 +158,7 @@ const Spin: React.FC<SpinClassProps> = (props) => {
     </div>
   );
 
-  if (isNestedPattern) {
+  if (isNestedPattern && !fullscreen) {
```
Comment (MadCcc) [inline components/spin/index.tsx:161]: `fullscreen` should be part of `isNestedPattern`, put it into `useMemo`?

**Comment by MadCcc** [comment]
> Hey @MadCcc, I fixed the tip usage, I just feel that I'm repeating code a little bit in css for this, what do you think?

@Rafael-Martins Yes, it should be only contain `color: token.colorTextLightSolid`, and other style should be reused.

**Review by MadCcc — APPROVED** [review]

**Comment by afc163** [comment]
Could we add a fade-in-out animation?

**Pull request merged**